### PR TITLE
Make some `newtype_index!` derived impls opt-in instead of opt-out

### PR DESCRIPTION
**Pull request by nnethercote**
Opt-in is the standard Rust way of doing things, and avoids some unnecessary dependencies on the `rustc_serialize` crate.

r? @lcnr

### CODE REVIEW

**Comment by fbstj** [comment]
nit: I think you got your PR title inverted

**Review by compiler-errors — APPROVED** [review]
r=me after rebased

**Comment by nnethercote** [comment]
I rebased.

@bors r=compiler-errors

**Comment by nnethercote** [comment]
I rebased again.

@bors r=compiler-errors

File: compiler/rustc_index_macros/src/lib.rs [hunk 1]
```diff
@@ -18,8 +18,19 @@ mod newtype;
...
+///
+/// Accepted attributes for customization:
+/// - #[derive(HashStable_Generic)]/#[derive(HashStable)]: derives
+///   `HashStable`, as normal.
+/// - #[encodable]: derives `Encodable`/`Decodable`.
+/// - #[orderable]: derives `PartialOrd`/`Ord`, plus step-related methods.
+/// - #[debug_format = "Foo({})"]: derives `Debug` with particular output.
+/// - #[max = 0xFFFF_FFFD]: specifies the max value, which allows niche
+///   optimizations. The default max value is 0xFFFF_FF00.
+/// - #[gate_rustc_only]: makes parts of the generated code nightly-only.
```
Comment (klensy) [inline compiler/rustc_index_macros/src/lib.rs:33]: https://github.com/rust-lang-ci/rust/actions/runs/6990882559/job/19020846394#step:26:27071
```
   warning: unresolved link to `encodable`
    --> compiler/rustc_index_macros/src/lib.rs:28:9
     |
  28 | /// - #[encodable]: derives `Encodable`/`Decodable`.
     |         ^^^^^^^^^ no item named `encodable` in scope
     |
     = help: to escape `[` and `]` characters, add '\' before them like `\[` or `\]`
     = note: `#[warn(rustdoc::broken_intra_doc_links)]` on by default
  
  warning: unresolved link to `orderable`
    --> compiler/rustc_index_macros/src/lib.rs:29:9
     |
  29 | /// - #[orderable]: derives `PartialOrd`/`Ord`, plus step-related methods.
     |         ^^^^^^^^^ no item named `orderable` in scope
     |
     = help: to escape `[` and `]` characters, add '\' before them like `\[` or `\]`
  
  warning: unresolved link to `gate_rustc_only`
    --> compiler/rustc_index_macros/src/lib.rs:33:9
     |
  33 | /// - #[gate_rustc_only]: makes parts of the generated code nightly-only.
     |         ^^^^^^^^^^^^^^^ no item named `gate_rustc_only` in scope
     |
     = help: to escape `[` and `]` characters, add '\' before them like `\[` or `\]`
```

Comment (nnethercote) [reply compiler/rustc_index_macros/src/lib.rs:33]: Should be fixed by #118338.

How did you notice this? I never notice warnings buried deep within CI runs like this...

Comment (klensy) [reply compiler/rustc_index_macros/src/lib.rs:33]: Reading logs from time to time. It's nice to have github action to automate log diffing and grepping to notice interesting stuff.

**Pull request merged**